File: leetcode/go/565.array-nesting.go
func arrayNesting(nums []int) int {
    n := len(nums)
    visited := make([]bool, n)
    ans := 0
    for i := range n {
        if !visited[i] {
            setLength := 0
            j := i
            for !visited[j] {
                visited[j] = true
                j = nums[j]
                setLength++
                ans = max(ans, setLength)
            }
        }
    }
    return ans
}

File: leetcode/rust-leetcode/src/565.array-nesting.rs
impl Solution {
    pub fn array_nesting(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut visited = vec![false; n];
        let mut ans = 0;
        for k in 0..n {
            if !visited[k] {
                let mut set_length = 0;
                let mut j = k;
                while !visited[j] {
                    visited[j] = true;
                    j = nums[j] as usize;
                    set_length += 1;
                    ans = ans.max(set_length);
                }
            }
        }
        ans
    }
}

